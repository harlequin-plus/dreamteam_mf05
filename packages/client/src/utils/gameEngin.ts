import { styles } from '../components/GameBoard/tempStyles'
import { GameMoves } from './gameMoves'
import { getColorForValue, keysMapper } from './helpers'

export class GameEngine extends GameMoves {
  constructor() {
    super()
  }

  public ctxRef: CanvasRenderingContext2D | null = null
  public completed = false
  public score = 0

  setScore = (score: number) => {
    this.score = score
    if (this.setScoreCallback) {
      this.setScoreCallback(score)
    }
  }

  start = (canvasRef: React.MutableRefObject<HTMLCanvasElement | null>) => {
    this.completed = false
    this.board = this.emptyBoard()
    this.setScore(0)

    if (canvasRef.current) {
      const canvas = canvasRef.current
      const ctx = canvas.getContext('2d')

      if (ctx) {
        this.ctxRef = ctx
        window.addEventListener('keydown', this.onKeyboardPress)
      }
    }

    this.drawBoard(this.ctxRef)
    this.drawTiles(this.ctxRef)
    this.addNewTile(this.ctxRef)
  }

  finish = () => {
    window.removeEventListener('keydown', this.onKeyboardPress)
  }

  onKeyboardPress = (event: KeyboardEvent) => {
    let moved = false

    // remove from gameEngie
    switch (keysMapper[event.key]) {
      case 'up':
        event.preventDefault()
        moved = this.moveUp()
        break
      case 'down':
        event.preventDefault()
        moved = this.moveDown()
        break
      case 'left':
        event.preventDefault()
        moved = this.moveLeft()
        break
      case 'right':
        event.preventDefault()
        moved = this.moveRight()
        break

      default:
        break
    }

    if (moved) {
      this.drawBoard(this.ctxRef)
      this.addNewTile(this.ctxRef)
    }

    this.checkFinish()
  }

  checkFinish = (): boolean => {
    let largestElement = Number.NEGATIVE_INFINITY

    for (const row of this.board) {
      for (const element of row) {
        if (element > largestElement) {
          largestElement = element
        }
      }
    }

    if (largestElement === 2048 && !this.completed) {
      alert(
        `Congrats! You are about to score 2048! Continue if you want bigger score`
      )

      this.completed = true
      return true
    }
    return false
  }

  addNewTile = (ctxRef: CanvasRenderingContext2D | null) => {
    const emptyTiles: { row: number; col: number }[] = []

    for (let i = 0; i < this.board.length; i++) {
      for (let j = 0; j < this.board[i].length; j++) {
        if (this.board[i][j] === 0) {
          emptyTiles.push({ row: i, col: j })
        }
      }
    }

    if (emptyTiles.length === 1) {
      const randomIndex = Math.floor(Math.random() * emptyTiles.length)
      const { row, col } = emptyTiles[randomIndex]

      this.board[row][col] = 2

      this.drawTiles(ctxRef)
      this.drawBoard(ctxRef)

      if (this.moveNotPossible()) {
        if (confirm(`No moves available. New game will be started`)) {
          this.setScore(0)
          for (let i = 0; i < this.board.length; i++) {
            for (let j = 0; j < this.board[i].length; j++) {
              this.board[i][j] = 0
            }
          }
          this.addNewTile(ctxRef)
          this.drawBoard(ctxRef)
          this.drawTiles(ctxRef)
        }
      }
    }

    if (emptyTiles.length > 0) {
      const randomIndex = Math.floor(Math.random() * emptyTiles.length)
      const { row, col } = emptyTiles[randomIndex]

      this.board[row][col] = 2
      this.drawBoard(ctxRef)
      this.drawTiles(ctxRef)
    }
  }
  drawBoard = (ctxRef: CanvasRenderingContext2D | null) => {
    if (ctxRef) {
      const ctx = ctxRef
      ctx.fillStyle = '#bbada0'
      ctx.fillRect(5, 5, 385, 385)

      for (let i = 0; i < this.board.length; i++) {
        for (let j = 0; j < this.board[i].length; j++) {
          ctx.fillStyle = getColorForValue(this.board[i][j])
          ctx.fillRect(10 + j * 95, 10 + i * 95, 90, 90)
        }
      }
    }
  }

  drawTiles = (ctxRef: CanvasRenderingContext2D | null) => {
    if (ctxRef) {
      const ctx = ctxRef
      ctx.fillStyle = styles.color
      ctx.font = styles.font
      ctx.textAlign = <CanvasTextAlign>styles.alignItems
      ctx.textBaseline = <CanvasTextBaseline>styles.textBaseline

      for (let i = 0; i < this.board.length; i++) {
        for (let j = 0; j < this.board[i].length; j++) {
          if (this.board[i][j] !== 0) {
            ctx.fillText(String(this.board[i][j]), 55 + j * 95, 55 + i * 95)
          }
        }
      }
    }
  }
}
