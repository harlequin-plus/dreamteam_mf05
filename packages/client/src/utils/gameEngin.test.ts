import { GameEngine } from './gameEngin'
import 'jest-canvas-mock'

describe('GameEngine', () => {
  let gameEngine: GameEngine
  let fullBoard: number[][]
  beforeEach(function () {
    gameEngine = new GameEngine()
    fullBoard = [
      [2, 4],
      [8, 16],
    ]
  })

  it(`moveUp должен вернуть False при заполненных ячейках`, function () {
    gameEngine.board = fullBoard
    const result = gameEngine.moveUp()
    expect(result).toBeFalsy()
  })
  it(`moveUp  должен переместить значения с нижнего ряда на верхний и сложить равные ячейки`, function () {
    gameEngine.board = [
      [0, 4],
      [4, 4],
    ]
    gameEngine.moveUp()
    expect(gameEngine.board).toEqual([
      [4, 8],
      [0, 0],
    ])
  })

  it(`moveDown должен вернуть False при заполненных ячейках`, function () {
    gameEngine.board = fullBoard
    const result = gameEngine.moveDown()
    expect(result).toBeFalsy()
  })
  it(`moveDown  должен переместить значения с верхнего ряда на нижний и сложить равные ячейки`, function () {
    gameEngine.board = [
      [2, 4],
      [0, 4],
    ]
    gameEngine.moveDown()
    expect(gameEngine.board).toEqual([
      [0, 0],
      [2, 8],
    ])
  })

  it(`moveRight должен вернуть False при заполненных ячейках`, function () {
    gameEngine.board = fullBoard
    const result = gameEngine.moveRight()
    expect(result).toBeFalsy()
  })
  it(`moveRight  должен переместить значения с левой части на правую и сложить равные ячейки`, function () {
    gameEngine.board = [
      [2, 0],
      [4, 4],
    ]
    gameEngine.moveRight()
    expect(gameEngine.board).toEqual([
      [0, 2],
      [0, 8],
    ])
  })

  it(`moveLeft должен вернуть False при заполненных ячейках`, function () {
    gameEngine.board = fullBoard
    const result = gameEngine.moveLeft()
    expect(result).toBeFalsy()
  })
  it(`moveLeft  должен переместить значения с правой части на левую и сложить равные ячейки`, function () {
    gameEngine.board = [
      [0, 2],
      [4, 4],
    ]
    gameEngine.moveLeft()
    expect(gameEngine.board).toEqual([
      [2, 0],
      [8, 0],
    ])
  })

  it(`moveNotPossible должен вернуть true при заполненных ячейках`, function () {
    gameEngine.board = fullBoard
    const result = gameEngine.moveNotPossible()
    expect(result).toBeTruthy()
  })
})

describe('drawBoard', () => {
  let canvas: HTMLCanvasElement, ctx: CanvasRenderingContext2D
  let ctxRef: React.MutableRefObject<CanvasRenderingContext2D>
  let gameEngine: GameEngine
  const board = [
    [2, 4],
    [8, 8],
  ]
  beforeEach(function () {
    canvas = document.createElement('canvas')
    ctx = canvas.getContext('2d') as CanvasRenderingContext2D
    ctxRef = { current: ctx }
    gameEngine = new GameEngine()
  })

  it(`должен вернуть снимок границы поля игры`, function () {
    gameEngine.board = board
    gameEngine.drawBoard(ctxRef.current)
    const events = ctxRef.current.__getEvents()

    expect(events).toMatchSnapshot()
  })
  it(`должен вернуть снимок ячеек поля игры`, function () {
    gameEngine.board = board
    gameEngine.drawTiles(ctxRef.current)
    const events = ctxRef.current.__getEvents()
    console.log(events)
    expect(events).toMatchSnapshot()
  })
})
