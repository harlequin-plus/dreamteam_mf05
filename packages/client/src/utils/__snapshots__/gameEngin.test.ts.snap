// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`drawBoard должен вернуть снимок границы поля игры 1`] = `
Array [
  Object {
    "props": Object {
      "value": "#bbada0",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "height": 390,
      "width": 390,
      "x": 5,
      "y": 5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "value": "#eee4da",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "height": 90,
      "width": 90,
      "x": 10,
      "y": 10,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "value": "#ede0c8",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "height": 90,
      "width": 90,
      "x": 105,
      "y": 10,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "value": "#f2b179",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "height": 90,
      "width": 90,
      "x": 10,
      "y": 105,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
  Object {
    "props": Object {
      "value": "#f2b179",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "height": 90,
      "width": 90,
      "x": 105,
      "y": 105,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillRect",
  },
]
`;

exports[`drawBoard должен вернуть снимок ячеек поля игры 1`] = `
Array [
  Object {
    "props": Object {
      "value": "#776e65",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "value": "bold 40px Arial",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "center",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "2",
      "x": 55,
      "y": 55,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "4",
      "x": 150,
      "y": 55,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "8",
      "x": 55,
      "y": 150,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "8",
      "x": 150,
      "y": 150,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
]
`;
